src/engraving/libmscore/scorefile.cpp

```diff
@@ -169,16 +169,27 @@ void Score::write(XmlWriter& xml, bool selectionOnly, compat::WriteScoreHook& ho
     }
 
     if (!systemObjectStaves.empty()) {
-        // write which staves currently have system objects above them
-        xml.startElement("SystemObjects");
+        bool saveSysObjStaves = false;
         for (Staff* s : systemObjectStaves) {
-            // TODO: when we add more granularity to system object display, construct this string per staff
-            String sysObjForStaff = u"barNumbers=\"false\"";
-            // for now, everything except bar numbers is shown on system object staves
-            // (also, the code to display bar numbers on system staves other than the first currently does not exist!)
-            xml.tag("Instance", { { "staffId", s->idx() + 1 }, { "barNumbers", "false" } });
+            if (s->idx() != mu::nidx) {
+                saveSysObjStaves = true;
+                break;
+            }
+        }
+        if (saveSysObjStaves) {
+            // write which staves currently have system objects above them
+            xml.startElement("SystemObjects");
+            for (Staff* s : systemObjectStaves) {
+                if (s->idx() != mu::nidx) {
```

Comment:
...and these ones are to make sure that non-existing staves don't get written to the file...

Reply:
Yep! We don't save zombie staves to the file because the file doesn't save the undo stack.

Reply:
Same question here actually

Reply:
It looks like an empty `SystemObjects` tag won't do any harm, and it is very unlikely to happen anyway (indeed it would happen only when something goes wrong). So I don't think it's worth the extra loop. But I'm happy with any decision :)

In any case, I think it would be nice to replace each `if (s->idx() != mu::nidx) {` with `IF_ASSERT_FAILED(s->idx() != mu::nidx) { continue; }`, to make clear that this is an unexpected scenario.

Reply:
fair enough, changed!